concordium-consensus/src/Concordium/KonsensusV1/Consensus/Timeout.hs

```diff
@@ -143,41 +143,38 @@ receiveTimeoutMessage tm@TimeoutMessage{tmBody = TimeoutMessageBody{..}} skovDat
-                            -- QC for the round, and the highest certified block is for a later round.
-                            -- We require the latter check because it is not necessarily guaranteed by
-                            -- the former, and if we are to use the timeout message to produce a
-                            -- timeout certificate and subsequently a block, then we must necessarily
-                            -- catch-up in order to do so. This is because the produced block must
-                            -- descend from a block in at least round @qcRound tmQuorumCertificate@,
-                            -- but the highest certified block is currently the best candidate.
-                            RecentBlock BlockPendingOrUnknown
-                                | qcRound tmQuorumCertificate
-                                    < skovData ^. roundStatus . rsHighestCertifiedBlock . to cbRound,
-                                  Just (QuorumCertificateCheckedWitness certEpoch) <-
-                                    skovData ^. roundExistingQuorumCertificate (qcRound tmQuorumCertificate) ->
-                                    -- Check whether we have already checked a QC for the @Round@
-                                    -- and @Epoch@ of the @tmQuorumCertificate@.
-                                    if qcEpoch tmQuorumCertificate == certEpoch
-                                        then received finInfo finalizationCommittee Absent
-                                        else return $ Rejected BadQCEpoch
-                                | otherwise ->
-                                    return CatchupRequired
+                    checkForDuplicate $ do
+                        -- Check whether we have already checked a QC for the round of the quorum certificate
+                        -- associated with the timeout message and that the qc round is higher (with respect to round number) than
+                        -- our highest certified block.
+                        let existingQC = skovData ^. roundExistingQuorumCertificate (qcRound tmQuorumCertificate)
+                            oldOrCurrentHighestCertifiedQCRound = qcRound tmQuorumCertificate <= skovData ^. roundStatus . rsHighestCertifiedBlock . to cbRound
+                        case (existingQC, oldOrCurrentHighestCertifiedQCRound) of
+                            (Just (QuorumCertificateCheckedWitness certEpoch), True) -> do
+                                -- The epochs must match otherwise we reject.
+                                if qcEpoch tmQuorumCertificate == certEpoch
+                                    then received finInfo finalizationCommittee Absent
+                                    else return $ Rejected BadQCEpoch
+                            -- If we have not already checked a QC for the round of the quorum certificate assoicated
+                            -- with the timeout message.
+                            _ -> do
```

Comment:
Does it ever make sense to log the distinction between the 3 different scenarios here?